src/uu/touch/src/touch.rs

```diff
@@ -240,110 +284,203 @@ pub fn uu_app() -> Command {
         )
 }
 
-// Determine the access and modification time
-fn determine_times(matches: &ArgMatches) -> UResult<(FileTime, FileTime)> {
-    match (
-        matches.get_one::<OsString>(options::sources::REFERENCE),
-        matches.get_one::<String>(options::sources::DATE),
-    ) {
-        (Some(reference), Some(date)) => {
-            let (atime, mtime) = stat(Path::new(&reference), !matches.get_flag(options::NO_DEREF))?;
-            let atime = filetime_to_datetime(&atime).ok_or_else(|| {
-                USimpleError::new(1, "Could not process the reference access time")
-            })?;
-            let mtime = filetime_to_datetime(&mtime).ok_or_else(|| {
-                USimpleError::new(1, "Could not process the reference modification time")
-            })?;
-            Ok((parse_date(atime, date)?, parse_date(mtime, date)?))
+/// Execute the touch command.
+///
+/// # Errors
+///
+/// Possible causes:
+/// - The user doesn't have permission to access the file
+/// - One of the directory components of the file path doesn't exist.
+pub fn touch(files: &[InputFile], opts: &Options) -> Result<(), TouchError> {
+    let (atime, mtime) = match &opts.source {
+        Source::Reference(reference) => {
+            let (atime, mtime) = stat(reference, !opts.no_deref)
+                .map_err(|e| TouchError::ReferenceFileInaccessible(reference.to_owned(), e))?;
+
+            (atime, mtime)
+        }
+        Source::Now => {
+            let now = datetime_to_filetime(&Local::now());
+            (now, now)
+        }
+        &Source::Timestamp(ts) => (ts, ts),
+    };
+
+    let (atime, mtime) = if let Some(date) = &opts.date {
+        (
+            parse_date(
+                filetime_to_datetime(&atime).ok_or_else(|| TouchError::InvalidFiletime(atime))?,
+                date,
+            )?,
+            parse_date(
+                filetime_to_datetime(&mtime).ok_or_else(|| TouchError::InvalidFiletime(mtime))?,
+                date,
+            )?,
+        )
+    } else {
+        (atime, mtime)
+    };
+
+    for (ind, file) in files.iter().enumerate() {
+        let (path, is_stdout) = match file {
+            InputFile::Stdout => (Cow::Owned(pathbuf_from_stdout()?), true),
+            InputFile::Path(path) => (Cow::Borrowed(path), false),
+        };
+        touch_file(&path, is_stdout, opts, atime, mtime).map_err(|e| {
+            TouchError::TouchFileError {
+                path: path.into_owned(),
+                index: ind,
+                error: e,
+            }
+        })?;
+    }
+
+    Ok(())
+}
+
+/// Touch a single file
```

Comment:
sorry but i was expecting a bit more than just repeating the function name :)

Reply:
My bad, added more now